Share code for creating a fill quad scene in postprocessing passes

Oletus opened the pull request:
This deduplicates a lot of code that used to be the same in a lot of postprocessing passes.

Oletus committed:
Share code for creating a fill quad scene in postprocessing passes

This deduplicates a lot of code that used to be the same in a lot of postprocessing passes.

mrdoob commented on examples/js/postprocessing/EffectComposer.js:
```diff
@@ -192,6 +192,22 @@ THREE.Pass = function () {
+	var fillQuad = {};
+
+	fillQuad.camera = new THREE.OrthographicCamera( - 1, 1, 1, - 1, 0, 1 );
+	fillQuad.scene = new THREE.Scene();
```
I think we don't even need a scene anymore, I think this should work:

```js
renderer.render( quad, camera );
```

Oletus replied:
Yep, it seems to, changed it now.

Oletus committed:
Don't initialize scene for postprocessing passes

The quad can be passed to renderer.render() directly without being a part of a scene.

Oletus committed:
Use createFillQuadScene also in OutlinePass

mrdoob commented on examples/js/postprocessing/EffectComposer.js:
```diff
@@ -192,6 +192,20 @@ THREE.Pass = function () {
+	fillQuad.camera = new THREE.OrthographicCamera( - 1, 1, 1, - 1, 0, 1 );
+
+	fillQuad.quad = new THREE.Mesh( new THREE.PlaneBufferGeometry( 2, 2 ), material );
+	fillQuad.quad.frustumCulled = false; // Avoid getting clipped
```
I think this is not needed either?

Oletus replied:
Probably not, I just thought there was some reason to have it since so many of the postprocessing passes used it.

WestLangley commented:
We have referred to this construct as a _full-screen quad_.

Why not use this pattern?

```js
THREE.Pass.FullScreenQuad = function( material ) {

	this.camera = new THREE.OrthographicCamera( - 1, 1, 1, - 1, 0, 1 );

	this.quad = new THREE.Mesh( new THREE.PlaneBufferGeometry( 2, 2 ), material );
	this.quad.frustumCulled = false;

};
```
and then
```js
this.fsQuad = new THREE.Pass.FullScreenQuad( this.material );
```

mrdoob commented:
Wouldn't it be possible to just create one camera and mesh in `EffectComposer` and reuse it?

Oletus committed:
Better naming and API for full-screen quads

Now the same geometry and camera are shared between full-screen quad instances. A mesh is still created for each full-screen quad instance since that avoids repeatedly setting the material on the mesh in cases where a pass only needs one material.

Oletus commented:
I added reusing the same geometry and camera and also cleaned up the API as suggested.

Mugen87 commented on examples/js/postprocessing/EffectComposer.js:
```diff
@@ -203,3 +203,39 @@ Object.assign( THREE.Pass.prototype, {
+} );
+
+THREE.Pass.FullScreenQuad._camera = new THREE.OrthographicCamera( - 1, 1, 1, - 1, 0, 1 );
+THREE.Pass.FullScreenQuad._geometry = new THREE.PlaneBufferGeometry( 2, 2 );
```
We should consider to put the entire `THREE.Pass` definition into a closure. In this way, it's not necessary to pollute the `FullScreenQuad` namespace with private variables.

Oletus replied:
I hid the _camera and _geometry in a closure now.

If this is expected it would be good if the code style notes would mention something about it.

Oletus committed:
Hide FullScreenQuad private variables in a closure

WestLangley commented on examples/js/postprocessing/EffectComposer.js:
```diff
@@ -203,3 +203,45 @@ Object.assign( THREE.Pass.prototype, {
+THREE.Pass.FullScreenQuad = (function() {
+
+	var _camera = new THREE.OrthographicCamera( - 1, 1, 1, - 1, 0, 1 );
+	var _geometry = new THREE.PlaneBufferGeometry( 2, 2 );
```
I don"t think the underscores are needed now.

Oletus replied:
Amended the last commit to remove the underscores from the variables in the closure.

Oletus commented:
Is there still something to be done to get this merged? As far as I can see I addressed all the feedback so far.

mrdoob commented:
Looks good!

mrdoob commented:
Thanks!

Oletus commented:
Thanks for merging!
